[GGML] Add GEMM and GEMV for Q4_0 and Q4_K quantization using ggml.

mnalewaj opened the pull request:
This PR adds GEMM and GEMV for Q4_0 and Q4_K quantization using ggml:
- adds ggml repo to subprojects
- adds standalone unit tests for cpu_backend
- adds Q4_0 quantization
- adds Q4_K quantization
- adds GEMM and GEMM using Q4_0 Q4_K
- adds flag ENABLE_GGML

**Self evaluation:**
1. Build test: [X]Passed [ ]Failed [ ]Skipped
2. Run test: [X]Passed [ ]Failed [ ]Skipped

skykongkong8 reviewed (approved):
Overall LGTM, but please apply patch

mnalewaj commented on nntrainer/tensor/cpu_backend/ggml_interface/ggml_interface.cpp:
```diff
@@ -10,7 +10,187 @@
+    int step_N = N / delta;
+    int step_C = delta;
+    int step_B = blocks_per_4_rows * sizeof(block_q4_K) * delta;
+#pragma omp parallel for collapse(1) num_threads(16)
+    for (int i = 0; i < step_N; i++) {
+      ::ggml_gemm_q4_K_8x8_q8_K(K, C + i * step_C, ldc, (char *)B + i * step_B,
+                                QA.data(), M, delta);
+    }
```
We are working on better multi-threading.
Better solution probably will be added in next PR.

EunjuYang commented:
I've tested this PR in a Linux environment, and it does not automatically apply the ggml patch, resulting in a compilation error. Could you add the patch application to the build script for linux environment?

EunjuYang commented:
What about adding `-Denable-ggml=true` in CI to verify that it is successfully supported?

skykongkong8 commented:
> What about adding `-Denable-ggml=true` in CI to verify that it is successfully supported?

I wonder if such option is really required in the first place.
ggml supports all `x86` `aarch64` and even `riscv` ( although their functions are used on `x86` in this PR), so how about just removing such option?

EunjuYang commented:
> > What about adding `-Denable-ggml=true` in CI to verify that it is successfully supported?
> 
> I wonder if such option is really required in the first place. ggml supports all `x86` `aarch64` and even `riscv` ( although their functions are used on `x86` in this PR), so how about just removing such option?

make sense!

mnalewaj commented:
> I've tested this PR in a Linux environment, and it does not automatically apply the ggml patch, resulting in a compilation error. Could you add the patch application to the build script for linux environment?

We checked the issue and probably the reason of that is too old version of meson (on 1.6.1 and above it works).
To update meson run:
apt purge meson
pip install meson

And to PATH add ${HOME}/.local/bin/

Additionally I added specific revision of ggml in ggml.wrap file.

mnalewaj commented:
> > What about adding `-Denable-ggml=true` in CI to verify that it is successfully supported?
> 
> I wonder if such option is really required in the first place. ggml supports all `x86` `aarch64` and even `riscv` ( although their functions are used on `x86` in this PR), so how about just removing such option?



> 

You are right - it is a good idea.
According to your suggestion I have tried to remove this flag.
Unfortunately I faced several problem from CI.
And some tasks was canceled - thus I expect additional several more issues.
I thing removing this flag is a good approach in the future PR. Update of meson is required and other fixes...
Now I suggest leave it as is.

List of CI problems what I mensioned:

1. meson_test (ubuntu-22.04, -Denable-fp16=true)
https://github.com/nnstreamer/nntrainer/actions/runs/14663470913/job/41152832422?pr=3118#step:11:1

Probably reason: too old version of meson.


2. Tizen GBS build on Ubuntu (x86_64, --define "unit_test_1")
https://github.com/nnstreamer/nntrainer/actions/runs/14663470917/job/41152832430?pr=3118

[  330s] 
[  330s] meson.build:468:18: ERROR: Automatic wrap-based subproject downloading is disabled

Probably reason: too old version of meson.

3. Android NDK build on Ubuntu for arm64-v8a
https://github.com/nnstreamer/nntrainer/actions/runs/14663470947/job/41152832398?pr=3118

  /home/runner/work/nntrainer/nntrainer/nntrainer/tensor/cpu_backend/ggml_interface/ggml_interface.cpp:13:10: fatal error: 'ggml-cpu-quants.h' file not found
  #include "ggml-cpu-quants.h"
           ^~~~~~~~~~~~~~~~~~~
  1 error generated.

I don't know the reason...

skykongkong8 commented:
> Automatic wrap-based subproject downloading is disabled

Clearly it seems there are multiple issues for that. Let's make this as a known issue and cleanup in the near future.

dkjung reviewed (approved):
LGTM

skykongkong8 commented:
> > I've tested this PR in a Linux environment, and it does not automatically apply the ggml patch, resulting in a compilation error. Could you add the patch application to the build script for linux environment?
> 
> We checked the issue and probably the reason of that is too old version of meson (on 1.6.1 and above it works). To update meson run: apt purge meson pip install meson
> 
> And to PATH add ${HOME}/.local/bin/
> 
> Additionally I added specific revision of ggml in ggml.wrap file.

I checked this issue. Thanks!👍

jijoongmoon reviewed (approved):
LGTM

djeong20 reviewed (changes requested):
I believe the Ubuntu Meson test should pass regardless of the Meson version. When I tested this pull request (PR) locally with Meson version 1.6.1, I encountered a compilation error that needs to be addressed. 

```
_ggml-cpu_ggml-cpu-quants.c.o -c ../subprojects/ggml/src/ggml-cpu/ggml-cpu-quants.c
/tmp/ccmiAq95.s: Assembler messages:
/tmp/ccmiAq95.s:361: Error: unsupported instruction `vpdpbusd'
/tmp/ccmiAq95.s:419: Error: unsupported instruction `vpdpbusd'
/tmp/ccmiAq95.s:493: Error: unsupported instruction `vpdpbusd'
/tmp/ccmiAq95.s:566: Error: unsupported instruction `vpdpbusd'
```

Additionally, it seems that further work is required for the Tizen and Android builds. Please refer to #2922 for more details.

djeong20 commented:
FYI, downloading wrap-based subproject is not an option for Tizen. One option is to have ggml source file zipped as tar.gz and use it. Also, missing header on Android mostly occurs when the project is not properly packaged and linked.

mnalewaj commented:
> I believe the Ubuntu Meson test should pass regardless of the Meson version. When I tested this pull request (PR) locally with Meson version 1.6.1, I encountered a compilation error that needs to be addressed.
> 
> ```
> _ggml-cpu_ggml-cpu-quants.c.o -c ../subprojects/ggml/src/ggml-cpu/ggml-cpu-quants.c
> /tmp/ccmiAq95.s: Assembler messages:
> /tmp/ccmiAq95.s:361: Error: unsupported instruction `vpdpbusd'
> /tmp/ccmiAq95.s:419: Error: unsupported instruction `vpdpbusd'
> /tmp/ccmiAq95.s:493: Error: unsupported instruction `vpdpbusd'
> /tmp/ccmiAq95.s:566: Error: unsupported instruction `vpdpbusd'
> ```
> 
> Additionally, it seems that further work is required for the Tizen and Android builds. Please refer to #2922 for more details.

It seems that the problem with Error: unsupported instruction `vpdpbusd' is not connected with meson version.
I locally used the same version (1.6.1) and I never encountered that problem.
Probably the problem is connected with your specific local setup.
I found some discussion about this problem:
https://github.com/microsoft/onnxruntime/issues/24275#issuecomment-2833281968
Maybe it will be useful in your case.
I am not advanced build system engineer.

Do you expect add ggml to Tizen and Android build in this PR?

djeong20 commented:
@mnalewaj, Please refer to #3143 to properly build and link the third-party project for the Android environment. Let me know if you need any help!

EunjuYang commented:
Please rebase this PR to pass CI for Windows.

mnalewaj commented:
> Please rebase this PR to pass CI for Windows.

Done

mnalewaj commented:
> @mnalewaj, Please refer to #3143 to properly build and link the third-party project for the Android environment. Let me know if you need any help!

Android build with ggml is ready - please review.

djeong20 commented:
First of all, I want to appreciate the hard work! You did a great job. 👍 

I confirmed that the Android compilation works; however, I couldn't verify if there are any runtime errors. Here are the steps to follow:

1. Please modify the `test/jni/Android.mk` file to create an executable for the newly added unit test (CPU backend).
2. Follow the [guide](https://github.com/nnstreamer/nntrainer/tree/main/tools) to run `unittest_nntrainer_cpu_backend` and check if the unit tests pass on an Android device.

If you don’t have access to an Android device, let me know once you complete step 1, and I can verify it for you.

Additionally, I completed the Tizen build on PR 3143. Please refer to that for the Tizen build setups.

djeong20 commented:
@mnalewaj I tried running the unit tests on Android, and I get the following errors

```
[arm64-v8a] Compile++      : unittest_nntrainer_cpu_backend <= unittest_nntrainer_cpu_backend.cpp
/home/donghyeon/nntrainer/test/jni/../unittest/unittest_nntrainer_cpu_backend.cpp:219:14: error: no member named 'quantize_q4_K' in namespace 'nntrainer'
  219 |   nntrainer::quantize_q4_K(rhs_ptr, (void *)offline_qWeight_ptr, K, N, nullptr);
      |   ~~~~~~~~~~~^
/home/donghyeon/nntrainer/test/jni/../unittest/unittest_nntrainer_cpu_backend.cpp:221:14: error: no member named 'dequantize_row_q4_K' in namespace 'nntrainer'
  221 |   nntrainer::dequantize_row_q4_K(offline_qWeight_ptr, rhs_ptr_tmp, K * N);
      |   ~~~~~~~~~~~^
/home/donghyeon/nntrainer/test/jni/../unittest/unittest_nntrainer_cpu_backend.cpp:255:14: error: no member named 'init_backend' in namespace 'nntrainer'
  255 |   nntrainer::init_backend();
      |   ~~~~~~~~~~~^
/home/donghyeon/nntrainer/test/jni/../unittest/unittest_nntrainer_cpu_backend.cpp:268:14: error: no member named 'quantize_q4_0' in namespace 'nntrainer'
  268 |   nntrainer::quantize_q4_0(weights, (void *)q4_0_offline_qWeight_ptr, N, K,
      |   ~~~~~~~~~~~^
/home/donghyeon/nntrainer/test/jni/../unittest/unittest_nntrainer_cpu_backend.cpp:274:14: error: no member named 'repack_q4_0_to_q4_0_8' in namespace 'nntrainer'
  274 |   nntrainer::repack_q4_0_to_q4_0_8(q4_0_repacked_qWeight.data(),
      |   ~~~~~~~~~~~^
/home/donghyeon/nntrainer/test/jni/../unittest/unittest_nntrainer_cpu_backend.cpp:283:3: error: no member named 'gemm_q4_0' in namespace 'nntrainer'; did you mean simply 'gemm_q4_0'?
  283 |   nntrainer::gemm_q4_0(M, N, K, activations, K,
      |   ^~~~~~~~~~~~~~~~~~~~
      |   gemm_q4_0
/home/donghyeon/nntrainer/nntrainer/tensor/cpu_backend/cpu_backend.h:617:13: note: 'gemm_q4_0' declared here
  617 | extern void gemm_q4_0(const unsigned int M, const unsigned int N,
      |             ^
/home/donghyeon/nntrainer/test/jni/../unittest/unittest_nntrainer_cpu_backend.cpp:310:14: error: no member named 'quantize_q4_K' in namespace 'nntrainer'
  310 |   nntrainer::quantize_q4_K(weights, (void *)offline_qWeight_ptr, N, K, nullptr);
      |   ~~~~~~~~~~~^
/home/donghyeon/nntrainer/test/jni/../unittest/unittest_nntrainer_cpu_backend.cpp:315:14: error: no member named 'repack_q4_K_to_q4_K_8' in namespace 'nntrainer'
  315 |   nntrainer::repack_q4_K_to_q4_K_8(repacked_qWeight.data(), offline_qWeight_ptr,
      |   ~~~~~~~~~~~^
8 errors generated.
```

I believe this error occurs because the functions that throw errors are only defined in `x86_compute_backend.h`, not in `arm_compute_backend.h`. Please correctly locate these functions in `cpu_backend.h` to utilize fallback.h.

mnalewaj commented:
> Please check this [branch](https://github.com/djeong20/nntrainer/tree/ggml_q4) to reproduce.

Thank you for your branch - I will check it and I will fix the problems.

mnalewaj commented:
> @mnalewaj I tried running the unit tests on Android, and I get the following errors
> 
> ```
> [arm64-v8a] Compile++      : unittest_nntrainer_cpu_backend <= unittest_nntrainer_cpu_backend.cpp
> /home/donghyeon/nntrainer/test/jni/../unittest/unittest_nntrainer_cpu_backend.cpp:219:14: error: no member named 'quantize_q4_K' in namespace 'nntrainer'
>   219 |   nntrainer::quantize_q4_K(rhs_ptr, (void *)offline_qWeight_ptr, K, N, nullptr);
>       |   ~~~~~~~~~~~^
> /home/donghyeon/nntrainer/test/jni/../unittest/unittest_nntrainer_cpu_backend.cpp:221:14: error: no member named 'dequantize_row_q4_K' in namespace 'nntrainer'
>   221 |   nntrainer::dequantize_row_q4_K(offline_qWeight_ptr, rhs_ptr_tmp, K * N);
>       |   ~~~~~~~~~~~^
> /home/donghyeon/nntrainer/test/jni/../unittest/unittest_nntrainer_cpu_backend.cpp:255:14: error: no member named 'init_backend' in namespace 'nntrainer'
>   255 |   nntrainer::init_backend();
>       |   ~~~~~~~~~~~^
> /home/donghyeon/nntrainer/test/jni/../unittest/unittest_nntrainer_cpu_backend.cpp:268:14: error: no member named 'quantize_q4_0' in namespace 'nntrainer'
>   268 |   nntrainer::quantize_q4_0(weights, (void *)q4_0_offline_qWeight_ptr, N, K,
>       |   ~~~~~~~~~~~^
> /home/donghyeon/nntrainer/test/jni/../unittest/unittest_nntrainer_cpu_backend.cpp:274:14: error: no member named 'repack_q4_0_to_q4_0_8' in namespace 'nntrainer'
>   274 |   nntrainer::repack_q4_0_to_q4_0_8(q4_0_repacked_qWeight.data(),
>       |   ~~~~~~~~~~~^
> /home/donghyeon/nntrainer/test/jni/../unittest/unittest_nntrainer_cpu_backend.cpp:283:3: error: no member named 'gemm_q4_0' in namespace 'nntrainer'; did you mean simply 'gemm_q4_0'?
>   283 |   nntrainer::gemm_q4_0(M, N, K, activations, K,
>       |   ^~~~~~~~~~~~~~~~~~~~
>       |   gemm_q4_0
> /home/donghyeon/nntrainer/nntrainer/tensor/cpu_backend/cpu_backend.h:617:13: note: 'gemm_q4_0' declared here
>   617 | extern void gemm_q4_0(const unsigned int M, const unsigned int N,
>       |             ^
> /home/donghyeon/nntrainer/test/jni/../unittest/unittest_nntrainer_cpu_backend.cpp:310:14: error: no member named 'quantize_q4_K' in namespace 'nntrainer'
>   310 |   nntrainer::quantize_q4_K(weights, (void *)offline_qWeight_ptr, N, K, nullptr);
>       |   ~~~~~~~~~~~^
> /home/donghyeon/nntrainer/test/jni/../unittest/unittest_nntrainer_cpu_backend.cpp:315:14: error: no member named 'repack_q4_K_to_q4_K_8' in namespace 'nntrainer'
>   315 |   nntrainer::repack_q4_K_to_q4_K_8(repacked_qWeight.data(), offline_qWeight_ptr,
>       |   ~~~~~~~~~~~^
> 8 errors generated.
> ```
> 
> I believe this error occurs because the functions that throw errors are only defined in `x86_compute_backend.h`, not in `arm_compute_backend.h`. Please correctly locate these functions in `cpu_backend.h` to utilize fallback.h.

It is done - I fixed building of unit tests for Android and I run it successfully on Android Device.

mnalewaj commented:
> First of all, I want to appreciate the hard work! You did a great job. 👍
> 
> I confirmed that the Android compilation works; however, I couldn't verify if there are any runtime errors. Here are the steps to follow:
> 
> 1. Please modify the `test/jni/Android.mk` file to create an executable for the newly added unit test (CPU backend).
> 2. Follow the [guide](https://github.com/nnstreamer/nntrainer/tree/main/tools) to run `unittest_nntrainer_cpu_backend` and check if the unit tests pass on an Android device.
> 
> If you don’t have access to an Android device, let me know once you complete step 1, and I can verify it for you.
> 
> Additionally, I completed the Tizen build on PR 3143. Please refer to that for the Tizen build setups.

Hello, I need help.
I tried to run Tizen build locally.
I used the following commands:
	echo "deb [trusted=yes] http://download.tizen.org/tools/latest-release/Ubuntu_22.04/ /" | sudo tee /etc/apt/sources.list.d/tizen.list
	sudo apt-get update && sudo apt-get install -y gbs
	cp .github/workflows/tizen.gbs.conf ~/.gbs.conf
	export GITHUB_OUTPUT="date=$(date +'%Y-%m-%d')"
	gbs build --skip-srcrpm --define "_skip_debug_rpm 1" -A x86_64 --define "unit_test 1"
Build was finished successfully.

After that I modified the file: packaging/nntrainer.spec
and run again the commands:
	export GITHUB_OUTPUT="date=$(date +'%Y-%m-%d')"
	gbs build --skip-srcrpm --define "_skip_debug_rpm 1" -A x86_64 --define "unit_test 1"

Unfortunately changes of packaging/nntrainer.spec is not affecting that build (it is building with previous version of that file).
Could you give me a hint how I can deal with it?
How I can clear some cache or something like that?

Additionally I removed directory: ~/GBS-ROOT (there are some cached files) but it have not caused using the modified file: packaging/nntrainer.spec

djeong20 commented:
> Hello, I need help. I tried to run Tizen build locally. I used the following commands: echo "deb [trusted=yes] http://download.tizen.org/tools/latest-release/Ubuntu_22.04/ /" | sudo tee /etc/apt/sources.list.d/tizen.list sudo apt-get update && sudo apt-get install -y gbs cp .github/workflows/tizen.gbs.conf ~/.gbs.conf export GITHUB_OUTPUT="date=$(date +'%Y-%m-%d')" gbs build --skip-srcrpm --define "_skip_debug_rpm 1" -A x86_64 --define "unit_test 1" Build was finished successfully.
> 
> After that I modified the file: packaging/nntrainer.spec and run again the commands: export GITHUB_OUTPUT="date=$(date +'%Y-%m-%d')" gbs build --skip-srcrpm --define "_skip_debug_rpm 1" -A x86_64 --define "unit_test 1"
> 
> Unfortunately changes of packaging/nntrainer.spec is not affecting that build (it is building with previous version of that file). Could you give me a hint how I can deal with it? How I can clear some cache or something like that?
> 
> Additionally I removed directory: ~/GBS-ROOT (there are some cached files) but it have not caused using the modified file: packaging/nntrainer.spec

Hello @mnalewaj, you need to add `--include-all` to reflect local changes. So it should be as follows.
 `gbs build --skip-srcrpm --define "_skip_debug_rpm 1" -A x86_64 --define "unit_test 1" --include-all`

mnalewaj commented:
Hello @djeong20,
I am afraid that I am not able to resolve problems with Tizen build in short time.
There are several ggml specific problem with building that module on Tizen.
But this PR is very important to further work (we have urgent task to do). Currently it is a blocker and many other PRs are waiting for it.
So I suggest disable ggml on Tizen and merge this PR to main.
This decision will save me and others working time.
Thus if you agree I will revert the changes enabling ggml for tizen and the PR will be ready to merge.
Building Tizen with ggml will be added in next PR.

mnalewaj commented:
Currently build on Tizen is almost done.
To do:
- removing some warnings during ggml building - currently flag 'werror=false' in the main meson.build file is necessary for successful building.
- changing of location output ggml libs for x86 - currently it is: nntrainer_libdir / 'arm64-v8a' for all builds

djeong20 commented:
Hi @mnalewaj, I really appreciate your hard work! This task is certainly challenging, and I understand the difficulties involved in ensuring everything builds correctly.

Please feel free to proceed as you suggested. Reverting the changes and merging the pull request (PR) seems like a reasonable approach given the current urgency. We can address the Tizen-specific issues in a follow-up PR.

Additionally, it would be helpful if you could document any known issues on GitHub Issues so we can track and resolve them accordingly.

djeong20 commented:
> Currently build on Tizen is almost done. To do:
> 
> * removing some warnings during ggml building - currently flag 'werror=false' in the main meson.build file is necessary for successful building.
> * changing of location output ggml libs for x86 - currently it is: nntrainer_libdir / 'arm64-v8a' for all builds

Hello @mnalewaj, the second issue seems to be because of the CMake install directory in [meson.build](https://github.com/mnalewaj/nntrainer_mnalewaj/blob/bea686568a7a90a8061b94effe5de68e6f942287/meson.build#L474C3-L474C91).

We can add a condition to install only when it is Android, and I believe the issue of installing at arm64-v8a for all builds can be resolved.

```
if get_option('platform') == 'android'
  ggml_options.add_cmake_defines({'CMAKE_INSTALL_LIBDIR': nntrainer_libdir / 'arm64-v8a'})
endif
```

skykongkong8 commented:
@mnalewaj Thanks for your hard work! All your works were merged with #3176 !
